### PR TITLE
MEN-1724: add deployment device count and filtering by date

### PR DESCRIPTION
**Pull request by GregorioDiStefano**

### CODE REVIEW

**Comment by GregorioDiStefano** [comment]
Preview before I start adding tests

File: resources/deployments/controller/controller_deployments.go [hunk 1]
```diff
@@ -15,15 +15,17 @@
...
-	"net/http"
-	"net/url"
-
 	"github.com/ant0ine/go-json-rest/rest"
 	"github.com/asaskevich/govalidator"
 	"github.com/mendersoftware/go-lib-micro/identity"
 	"github.com/mendersoftware/go-lib-micro/log"
 	"github.com/mendersoftware/go-lib-micro/rest_utils"
 	"github.com/pkg/errors"
+	"net/http"
```
Comment (kjaskiewiczz) [inline resources/deployments/controller/controller_deployments.go]: can you use previous import formatting (goimports like)?

File: resources/deployments/controller/controller_deployments.go [hunk 2]
```diff
@@ -341,6 +369,18 @@ func (d *DeploymentsController) LookupDeployment(w rest.ResponseWriter, r *rest.
...
+	for idx, dep := range deps {
+		//REVIEW: should I return here on error?
+
+		deviceDeployments, err := d.model.GetDeviceStatusesForDeployment(ctx, *dep.Id)
+		if err != nil {
+			d.view.RenderError(w, r, err, http.StatusBadRequest, l)
+			return
+		}
+
+		deps[idx].Count = len(deviceDeployments)
```
Comment (kjaskiewiczz) [inline resources/deployments/controller/controller_deployments.go]: maybe there is a better way of counting objects then retrieving all of them (there is a count in mongo, but I don't know how it works)

**Comment by GregorioDiStefano** [comment]
To summarize the changes: 

1. the deployments/deployments/ endpoint has two new query parameters: created_after, and created_before, both which take a UTC epoch timestamp and filter the deployment history accordingly. 

2.  a device_count is added to every deployment listening, so you know how many devices were targeted with the deployment. 

3. an internal api is added, when you can specify a tenant_id and, and get it's deployment history. This requires a change in the go-lib-micro, that is simple, but I haven't pushed yet in case there is a better way of doing this

File: resources/tenants/controller/tenants.go [hunk 3]
```diff
@@ -15,25 +15,37 @@
...
+	"net/http"
 )
 
 type Controller struct {
 	model model.Model
 }
 
+type DeploymentsControllerWithTenant struct {
+	model                 model.Model
+	deploymentsController deploymentsController.DeploymentsController
```
Comment (kjaskiewiczz) [inline resources/tenants/controller/tenants.go]: tenant controller with deployments controller?
is this the only way?
it looks really weird to me

Comment (GregorioDiStefano) [reply resources/tenants/controller/tenants.go]: I can give this a shot:

https://docs.mongodb.com/manual/reference/method/db.collection.count/

File: resources/deployments/controller/controller_deployments.go (hunk 2 above)
Comment (GregorioDiStefano) [reply resources/deployments/controller/controller_deployments.go]: can give this a shot:

https://docs.mongodb.com/manual/reference/method/db.collection.count/

File: resources/deployments/controller/controller_deployments.go [hunk 4]
```diff
@@ -297,6 +299,24 @@ func ParseLookupQuery(vals url.Values) (deployments.Query, error) {
 		query.SearchText = search
 	}
 
+	createdBefore := vals.Get("created_before")
+	if createdBefore != "" {
+		if createdBeforeTime, err := parseEpochToTimestamp(createdBefore); err != nil {
+			return query, err
```
Comment (maciejmrowiec) [inline resources/deployments/controller/controller_deployments.go]: error message won't contain from which param issue comes from, use errors.Wrap to deal with messaging

File: resources/deployments/controller/controller_deployments.go [hunk 5]
```diff
@@ -341,6 +369,18 @@ func (d *DeploymentsController) LookupDeployment(w rest.ResponseWriter, r *rest.
 		return
 	}
 
+	for idx, dep := range deps {
```
Comment (maciejmrowiec) [inline resources/deployments/controller/controller_deployments.go]: This is business logic, should be part of business model. Controller handles protocol logic.

File: resources/deployments/controller/controller_deployments.go (hunk 2 above)
Comment (maciejmrowiec) [reply resources/deployments/controller/controller_deployments.go]: Don't count it ourself, use count query.

Comment (maciejmrowiec) [reply resources/deployments/controller/controller_deployments.go]: Alternatively, cache the count then creating deployment object in the first place. This way you avoid having to extra query

File: resources/deployments/deployment.go [hunk 6]
```diff
@@ -80,6 +80,9 @@ type Deployment struct {
 	// Initialized with the "pending" counter set to total device count for deployment.
 	// Individual counter incremented/decremented according to device status updates.
 	Stats map[string]int `json:"-"`
+
+	// Total number of devices targeted
+	Count int `json:"device_count,omitempty"`
```
Comment (maciejmrowiec) [inline resources/deployments/deployment.go]: This field is going to be stored in database, even if in current implementation is only alias for another query in response. Should be excluded from being stored  `bson:"-"`

Comment (maciejmrowiec) [reply resources/deployments/deployment.go]: Call variable DeviceCount, Count is not self explanatory  on it's own.

File: resources/tenants/controller/tenants.go (hunk 3 above)
Comment (maciejmrowiec) [reply resources/tenants/controller/tenants.go]: shouldn't that be model ?

File: resources/tenants/controller/tenants.go [hunk 7]
```diff
@@ -54,3 +66,23 @@ func (c *Controller) ProvisionTenantsHandler(w rest.ResponseWriter, r *rest.Requ
...
+		rest_utils.RestErrWithLog(w, r, l, nil, http.StatusBadRequest)
+	}
+
+	//REVIEW: WithoutAuthContext would need to be added to go-lib-micro
+	tenantCtx := identity.WithoutAuthContext(ctx, &identity.Identity{
+		Tenant: tenantID,
+	})
+
+	r.Request = r.WithContext(tenantCtx)
+	c.deploymentsController.LookupDeployment(w, r)
```
Comment (maciejmrowiec) [inline resources/tenants/controller/tenants.go]: share model instead model (once business logic is cleaned up and moved down the stack)

**Comment by maciejmrowiec** [comment]
API spec missing

File: resources/tenants/controller/tenants.go (hunk 7 above)
Comment (GregorioDiStefano) [reply resources/tenants/controller/tenants.go]: but the controller is already doing all the query parsing, I'd have to re-implement it here .

**Comment by GregorioDiStefano** [comment]
@maciejmrowiec can you take a look?
I think adding an acceptance test for the internal API makes sense, and I'll open a separate task.

File: resources/deployments/controller/controller_deployments.go [hunk 8]
```diff
@@ -297,6 +298,24 @@ func ParseLookupQuery(vals url.Values) (deployments.Query, error) {
 		query.SearchText = search
 	}
 
+	createdBefore := vals.Get("created_before")
+	if createdBefore != "" {
+		if createdBeforeTime, err := parseEpochToTimestamp(createdBefore); err != nil {
+			return query, errors.Wrap(err, "timestamp parsing failed")
```
Comment (maciejmrowiec) [inline resources/deployments/controller/controller_deployments.go]: include which query param are you parsing in error message, if you use both in request you'll need to know which is wrong

File: docs/internal_api.yml [hunk 9]
```diff
@@ -99,6 +99,73 @@ paths:
...
+        - name: page
+          in: query
+          description: Results page number
+          required: false
+          type: number
+          format: integer
+          default: 1
+        - name: created_before
+          in: query
+          description: List only deployments created before Unix timestamp (UTC)
```
Comment (maciejmrowiec) [inline docs/internal_api.yml]: db queries include greater or equal and less than equal, include that in docs

File: resources/tenants/controller/tenants.go [hunk 10]
```diff
@@ -18,22 +18,34 @@ import (
...
+	model deploymentsModel.DeploymentsModel
+}
+
 func NewController(model model.Model) *Controller {
 	return &Controller{
 		model: model,
 	}
 }
 
+func NewDeploymentsTenantController(model deploymentsModel.DeploymentsModel) *TenantDeploymentController {
```
Comment (maciejmrowiec) [inline resources/tenants/controller/tenants.go]: Why not use existing Controller instead adding another one.

Comment (GregorioDiStefano) [reply resources/tenants/controller/tenants.go]: It's not possible to create a new method in a non-local package (I cant create a method of deploymentsController inside the 'tenant' package)

Comment (maciejmrowiec) [reply resources/tenants/controller/tenants.go]: Controller is in same package, it's in the same file.

Comment (GregorioDiStefano) [reply resources/tenants/controller/tenants.go]: right, I see what you mean now, and fixed it.

File: resources/tenants/controller/tenants.go [hunk 11]
```diff
@@ -34,6 +37,12 @@ func NewController(model model.Model) *Controller {
 	}
 }
 
+func NewDeploymentController(depsModel *deploymentsModel.DeploymentsModel) *Controller {
```
Comment (maciejmrowiec) [inline resources/tenants/controller/tenants.go]: There is already New function for this structure, it should handle all the init handling.
Depending on which was used to create object some methods are functional and some not is asking for trouble. When creating Controller instance all methods should be usable.

Make NewController take both models as configuration.

File: routing.go [hunk 12]
```diff
@@ -247,3 +250,13 @@ func NewTenantsResourceRoutes(controller *tenantsController.Controller) []*rest.
 		rest.Post(ApiUrlInternal+"/tenants", controller.ProvisionTenantsHandler),
 	}
 }
+
+func NewTenantDeploymentsRoutes(controller *tenantsController.Controller) []*rest.Route {
```
Comment (maciejmrowiec) [inline routing.go]: There is already router for `tenantsController` no need for yet another one.
Made sense if there was different controller around, but when using one then it's just extra code.

File: routing.go [hunk 13]
```diff
@@ -162,17 +162,21 @@ func NewRouter(c config.ConfigReader) (rest.App, error) {
 		new(deploymentsView.DeploymentsView))
 	limitsController := limitsController.NewLimitsController(limitsModel,
 		new(view.RESTView))
+
+	tenantsDeploymentController := tenantsController.NewDeploymentController(deploymentModel)
```
Comment (maciejmrowiec) [inline routing.go]: there is already instance of this object available `tenantsController`,  no need for another one

File: docs/internal_api.yml [hunk 14]
```diff
@@ -99,6 +99,73 @@ paths:
           description: Internal server error.
           schema:
            $ref: "#/definitions/Error"
+  /deployments/{id}/deployments:
```
Comment (maciejmrowiec) [inline docs/internal_api.yml]: did you mean `tenants/{id}....`?

Comment (GregorioDiStefano) [reply docs/internal_api.yml]: no, I didn't. 

The end point is: api/internal/v1/deployments/<id>/deployments

File: routing.go [hunk 15]
```diff
@@ -237,13 +238,13 @@ func NewLimitsResourceRoutes(controller *limitsController.LimitsController) []*r
...
-func NewTenantsResourceRoutes(controller *tenantsController.Controller) []*rest.Route {
-
+func TenantRoutes(controller *tenantsController.Controller) []*rest.Route {
 	if controller == nil {
 		return []*rest.Route{}
 	}
 
 	return []*rest.Route{
 		rest.Post(ApiUrlInternal+"/tenants", controller.ProvisionTenantsHandler),
+		rest.Get(ApiUrlInternal+"/:tenant/deployments", controller.DeploymentsPerTenantHandler),
```
Comment (maciejmrowiec) [inline routing.go]: route looks a bit off, should that be something like:

internalAPI `/tenants/:tenant/deployments`

**Comment by GregorioDiStefano** [comment]
@maciejmrowiec i think this merge-able now

**Pull request merged**